[Entry] Algorithm: Adding kadane.md into algorithm folder (#2614)
Codecademy/docs: content/general/concepts/algorithm/terms/euclidean-algorithm/kadane-algorithm.md

---
Title: 'Kadane's Algorithm'
Description: 'Returns the maximum sum of a subarray within an array of numbers'
Subjects:
  - 'Computer Science'
Tags:
  - 'Algorithms'
  - 'Arithmetic'
  - 'Arrays'
CatalogContent:
  - 'learn-java'
  - 'paths/computer-science'
---

> **`Kadane's algorithm`** is a dynamic programming approach utilized to identify the maximum sum of a subarray within a given array of numbers. The algorithm solves this problem efficiently in a single pass through the array by maintaining two variables: `maxEndingHere` and `maxSoFar`. The algorithm iterates through the array, updating these variables at each step to track the maximum subarray sum ending at the current element and the overall maximum subarray sum found so far. The final value of maxSoFar represents the maximum subarray sum in the given `array`. This algorithm guarantees a time complexity of `O(n)`, making it an optimal solution for this problem.

Reviewer: ```suggestion
**`Kadane's algorithm`** is a dynamic programming approach to efficiently finding the maximum sum of a subarray in a given array of numbers. The algorithm works as follows:
1. Initialize a current sum (variable called `maxEndingHere`) equal to the value of the element at the first position in the array (`arr[0]`).
2. Iterate through the array. At every position, set `maxEndingHere` to the maximum of the following two values: `maxEndingHere + arr[i]` or `arr[i]`, where `i` is the current position in the array.
3. Keep track of the greatest `maxEndingHere` encountered (ex: in a variable called `maxSoFar`).
4. Return `maxSoFar`.
This algorithm guarantees a time complexity of `O(n)`, making it an optimal solution for this problem.
```

## Explanation

**Here's how Kadane's algorithm can be applied to a problem:**

> - Start with the first number and add it to a running total. Then, move to the next number and determine whether to continue with the current sum or start a new sum from this number. Always aim to maintain the largest sum seen so far.

Reviewer: If you accept my suggestion from above, you can delete this line.

> - For example, consider the array {-2, -1, -3, 4, -1, 2, 1, -5, 4}. Initialize two integer variables: one named `maxEndingHere` and the other called `maxSoFar`. These variables serve to track the maximum subarray sum ending at the current position and the overall maximum subarray sum.

Reviewer: Put this in a section called "example". Use the ## Markdown header for this.

> - The loop commence with the first element, which is -2. So, both `maxEndingHere` and `maxSoFar` start at -2.

Reviewer: ```suggestion
The loop commences with the first element, which is -2. Therefore, both `maxEndingHere` and `maxSoFar` are initially -2.
```

> - As the loop proceed to the next element, which is -1, compare it to the sum of the current element and the previous value of `maxEndingHere`. Update `maxEndingHere` to **-1** because it's greater than the sum of **-2** and **-1**. Also, compare `maxSoFar` with the updated value of `maxEndingHere`. Since **-2** is less than **-1**, update `maxSoFar` to **-1**.

Reviewer: ```suggestion
Proceed to the next element, -1, and compare it to the sum of itself and the previous value of `maxEndingHere`.  Since -1 is greater than the sum of -2 and -1, set`maxEndingHere` to  -1. Also, compare `maxSoFar` with the updated value of `maxEndingHere`. Since -2 is less than -1 set `maxSoFar` to  -1.
```

---
Title: 'Kadane's Algorithm'
Description: 'Returns the maximum sum of a subarray within an array of numbers'
Subjects:
  - 'Computer Science'
Tags:
  - 'Algorithms'
  - 'Arithmetic'
  - 'Arrays'
CatalogContent:
  - 'learn-java'
  - 'paths/computer-science'
---

> Kadane's algorithm is an efficient way to find the maximum sum of a contiguous subarray within a given array of numbers. It iterates through the array and keeps track of the current maximum sum and the maximum sum found so far. By comparing the current element with the running sum, it decides whether to continue the subarray or start a new one. The algorithm guarantees a time complexity of O(n), making it an optimal solution for this problem.

Reviewer: This actually isn't a very clear explanation of Kadane's algorithm. Can you reword it?

## Explanation

> In problems related to `Kadane's algorithm` an array of numbers will be given, and the goal is to find the biggest sum that can be made using a bunch of consecutive numbers. Skipping any numbers or using them out of order is not allowed.

Reviewer: This statement repeats what was said in the previous section. Can you reword it or remove it entirely?

**Here's how it can be done:**

> - Start with the first number and add it to a running total. Then, move to the next number and decide whether it's better to continue with the current sum or start a new sum from this number. Always aim to keep the largest sum seen so far.

Reviewer: The key point that may not be so intuitive is what happens when the we are better off by starting a new sum from the current position. Can you include this in the example? I.e., show a case where it is actually better off by starting a new sum from the current position.
